src/charm.py

```diff
@@ -31,12 +31,53 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.install, self._on_install)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.mongodb_relation_joined, self._mongodb_relation_joined)
         self.framework.observe(self.on.start, self._on_start)
 
+    def _mongodb_relation_joined(self, event) -> None:
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self.app.planned_units() == len(self._peers.units) + 1:
+            logger.debug(
+                "planned units does not match joined units: %s != %s, will "
+                + "wait to initialise replica set until all planned units "
+                + "have joined",
+                str(self.app.planned_units()),
+                len(self.model.get_relation(PEER).units) + 1,
+            )
+            return
```

Comment:
That's what I initially thought too! But actually having a defer here leads to a forever wait. For example:

Say you have 3 units (`A`,`B`,`C`) (and `A` is the leader). When the leader observes the first `relation-joined` event this is because **one** of the units `B` joined, **but `C` has not joined** thus:

`self.app.planned_units() == len(self._peers.units) + 1` = `3 = 2` 

so the condition is not satisfied. If we defer we will **keep waiting**, and we will **not** observe `C` joining. In order to observe `C` we will need to wait for the **second** `relation-joined` event. Once the second `relation-joined` even fires we will observe `C` and: 

`self.app.planned_units() == len(self._peers.units) + 1` = `3 = 3`

And then will will be ready.

TL;DR: We cannot defer since we need to allow all `relation-joined` events to fire so we can observe all the peers. 

I hope that makes sense 😅 its kind of confusing, I'm also super happy to talk about this on MM. This was something that tripped me up for _awhile_

Reply:
@marceloneppel

Reply:
I see. Since the deferred event is in the head of the queue, it won't update `self.model.get_relation(PEER)`, is that correct?

Reply:
Yes exactly!